### PR TITLE
Allegra translation

### PR DESCRIPTION
**Pull request by nc6**
This fills out the remaining translation utilities Shelley->Allegra.

Individual commits have more detailed comments.

### CODE REVIEW

File: shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs [hunk 1]
```diff
@@ -3,8 +3,16 @@
...
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
 {-# LANGUAGE TypeFamilies #-}
 {-# LANGUAGE UndecidableInstances #-}
+{- We disable warnings for name shadowing because of
+https://gitlab.haskell.org/ghc/ghc/-/issues/14630, which means that we get
+shadowing warnings for the named field puns when used with a pattern synonym.
+-}
+{-# OPTIONS_GHC -Wno-name-shadowing #-}
```
Comment (mrBliss) [inline shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs:15]: Try the `DisambiguateRecordFields` extension, that often helps in these situations.

File: shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs [hunk 2]
```diff
@@ -203,13 +253,14 @@ instance Crypto c => TranslateEra (AllegraEra c) PPUPState where
         }
 
 instance Crypto c => TranslateEra (AllegraEra c) TxId where
-  translateEra _ = error "TODO Shelley to Allegra translation"
+  translateEra _ = pure . coerce
```
Comment (mrBliss) [inline shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs]: There is a default instance using `Coercible` (see the docstring of `TranslateEra`), I think you could use that.

File: shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs [hunk 3]
```diff
@@ -203,13 +253,14 @@ instance Crypto c => TranslateEra (AllegraEra c) PPUPState where
...
 
 instance Crypto c => TranslateEra (AllegraEra c) TxIn where
   translateEra ctxt (TxIn txid ix) = return $ TxIn (translateEra' ctxt txid) ix
 
 instance Crypto c => TranslateEra (AllegraEra c) TxOut where
-  translateEra _ = error "TODO Shelley to Allegra translation"
+  translateEra () (TxOutCompact addr cfval) =
```
Comment (mrBliss) [inline shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs:261]: Maybe you can use the default instance here too.

File: shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs [hunk 4]
```diff
@@ -290,3 +341,54 @@ instance Crypto c => TranslateEra (AllegraEra c) EpochState where
...
+      BootstrapWitness
+        { bwKey = bwKey,
+          bwSig = coerce bwSig,
+          bwChainCode,
+          bwAttributes
+        }
+
+instance
+  (Crypto c, Typeable kr) =>
+  TranslateEra (AllegraEra c) (WitVKey kr)
```
Comment (mrBliss) [inline shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs:368]: Same?

File: shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs [hunk 5]
```diff
@@ -290,3 +341,54 @@ instance Crypto c => TranslateEra (AllegraEra c) EpochState where
...
+  translateEra ctx (RegPool pp) = pure . RegPool $ translateEra' ctx pp
+  translateEra _ (RetirePool kh e) = pure $ RetirePool (coerce kh) e
+
+instance Crypto c => TranslateEra (AllegraEra c) MIRCert where
+  translateEra _ (MIRCert pot rewards) =
+    pure $
+      MIRCert pot (Map.mapKeysMonotonic coerce rewards)
+
+instance Crypto c => TranslateEra (AllegraEra c) Update where
+  translateEra _ (Update pp en) = pure $ Update (coerce pp) en
```
Comment (mrBliss) [inline shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs:393]: Same?

File: shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs (hunk 2 above)
Comment (nc6) [reply shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs]: Ah, yes, I saw that (and removed a bunch of these) but clearly missed this one.

File: shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs (hunk 3 above)
Comment (nc6) [reply shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs:261]: Alas, no

File: shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs (hunk 1 above)
Comment (nc6) [reply shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs:15]: But not this one, alas

File: shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs (hunk 5 above)
Comment (nc6) [reply shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs:393]: Nope :-(

File: shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs (hunk 4 above)
Comment (nc6) [reply shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs:368]: Nope :-(

**Review by JaredCorduan — APPROVED** [review]
LGTM

File: shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs [hunk 6]
```diff
@@ -56,9 +74,43 @@ instance Crypto c => TranslateEra (AllegraEra c) NewEpochState where
           nesPd = nesPd nes
         }
 
-instance Crypto c => TranslateEra (AllegraEra c) Tx where
+instance forall c. Crypto c => TranslateEra (AllegraEra c) Tx where
   type TranslationError (AllegraEra c) Tx = ()
-  translateEra _ = error "TODO Shelley to Allegra translation"
+  translateEra ctx (Tx body witness md) =
```
Comment (mrBliss) [inline shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs:79]: Are we guaranteed that the `TxId` are the same?

Comment (JaredCorduan) [reply shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs:79]: That is the intention of https://github.com/input-output-hk/cardano-ledger-specs/pull/1954 

obviously we'll need good tests!

Comment (nc6) [reply shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs:79]: For these eras, yes. The `TxId` type does not change, and the hash (which is in any case just a phantom) is fixed now.

**Pull request merged**